llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -6837,6 +6837,25 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
     Changed = true;
   }
 
+  {
+    Value *X, *Y;
+    const APInt *C;
+    ICmpInst::Predicate Pred = I.getPredicate();
+    if ((Pred == ICmpInst::ICMP_SGE || Pred == ICmpInst::ICMP_SLT) &&
+        match(Op0, m_OneUse(m_SRem(m_Value(X), m_Value(Y)))) &&
+        ((match(Op1, m_OneUse(m_c_Add(m_Deferred(Y), m_AllOnes()))) &&
```

Comment:
Could you please also handle negative divisors as @nikic suggested?
https://alive2.llvm.org/ce/z/AJq3Ch

Reply:
how can we apply this optimization with negative divisors?

Reply:
We can simplify `icmp sle/sgt (srem X, Y), (Y + 1)` into `icmp eq/ne (srem X, Y), (Y + 1)` if we know `Y` is negative.